### DISCUSSION TITLE
First io_uring_get_sqe Returns NULL

### DISCUSSION THREAD

**Discussion by nibanks** [axboe/liburing#632] [Q&A] [answered]
I'm trying to put together a project (https://github.com/nibanks/eventq) to compare different queue mechanism across platforms (IOCP, epoll, kqueue, io_uring) and I'm having problems figuring out io_uring. It seems if I [initialize the queue with a depth more than 1](https://github.com/nibanks/eventq/blob/main/eventq.h#L203), then `io_uring_get_sqe` returns `NULL`. Essentially, if I call the following two functions:

```c
bool eventq_initialize(eventq* queue) {
    return 0 == io_uring_queue_init(1, queue, 0);
}

void eventq_enqueue(eventq queue, eventq_sqe* sqe, uint32_t type, void* user_data, uint32_t status) {
    sqe->type = type;
    sqe->user_data = user_data;
    sqe->status = status;
    struct io_uring_sqe *io_sqe = io_uring_get_sqe(&queue);
    if (io_sqe == NULL) {
        printf("io_uring_get_sqe returned NULL\n"); // <---- This is printed out if > 1 is passed to io_uring_queue_init.
        return;
    }
    io_uring_prep_nop(io_sqe);
    io_uring_sqe_set_data(io_sqe, sqe);
    io_uring_submit(&queue); // TODO - Extract to separate function?
}
```
If I can get this working, we might integrate into Windows (my team owns the Windows networking stack and APIs), https://github.com/microsoft/msquic, and/or https://github.com/microsoft/xdp-for-windows. So I'd really like to figure this out.

**Comment by axboe** [comment, marked as answer]
Do you have a full example that shows this issue? Not something I'd seen before, and I'm 99.99% sure it's a bug in the app. But can't really say much with the code excerpt.

**Reply by nibanks** [reply]
Sorry for the noise, but because of the abstraction layer (across the different examples) I was passing the `struct io_uring` into the function directly instead of `struct io_uring*`. After fixing that (thanks @mtfriesen for pointing that out), now it queues properly, but my dequeue function just hangs.

I can ask a separate question for that, but am I doing this wrong? `UINT32_MAX` means to "wait forever":
```c
uint32_t eventq_dequeue(eventq* queue, eventq_cqe* events, uint32_t count, uint32_t wait_time) {
    if (wait_time != UINT32_MAX) {
        struct __kernel_timespec timeout;
        timeout.tv_sec += (wait_time / 1000);
        timeout.tv_nsec += ((wait_time % 1000) * 1000000);
        return io_uring_wait_cqes(queue, events, count, &timeout, 0);
    }
    return io_uring_wait_cqes(queue, events, count, 0, 0);
}
```

BTW, full example:
```
git clone https://github.com/nibanks/eventq.git
cd eventq && mkdir build && cd build
cmake -G 'Unix Makefiles' -DUSE_IO_URING=on ..
cmake --build .
./eventq
// HANG!
```

**Reply by nibanks** [reply]
Ok. I've made some progress. Apparently `io_uring_wait_cqes` waits for ALL events. I assumed it worked like other interfaces that would return the number of events/entries it returned. Is there an io_uring API like that?

**Reply by axboe** [reply]
It waits for as many events as you ask for. If you ask for min_events == 4, it'll block until 4 are available or your timeout is met. Again, I'm very sure this is an application issue and it's really hard to help with just snippets that don't really show everything that is going on.

**Reply by axboe** [reply]
If you just want to wait until at least 1 event is available, just pass in 1.

**Reply by nibanks** [reply]
Ok. Thanks. That helps!

**Reply by axboe** [reply]
The reason it doesn't return the number of events available is that that would be racy, as it can change between the wait and when you iterate them. Since the ring is shared between the kernel and the application, the number available is readily available by just looking at the cq ring. Normally you would not even do that, you'd just loop over available events and handle them, or return that number up the stack.

**Reply by axboe** [reply]
Also consider that if you do both submit and wait in one syscall, you really cannot return both the number submitted and the number waited on in a sane fashion. That's another reason that event waiting will return 0 for success, or < 0 for errors.
